// @angular/build/src/builders/extract-i18n/schema.js
"use strict";
// THIS FILE IS AUTOMATICALLY GENERATED. TO UPDATE THIS FILE YOU NEED TO CHANGE THE
// CORRESPONDING JSON SCHEMA FILE, THEN RUN devkit-admin build (or bazel build ...).
Object.defineProperty(exports, "__esModule", { value: true });
exports.I18NDuplicateTranslation = exports.Format = void 0;
/**
 * Output format for the generated file.
 */
var Format;
(function (Format) {
    Format["Arb"] = "arb";
    Format["Json"] = "json";
    Format["LegacyMigrate"] = "legacy-migrate";
    Format["Xlf"] = "xlf";
    Format["Xlf2"] = "xlf2";
    Format["Xlif"] = "xlif";
    Format["Xliff"] = "xliff";
    Format["Xliff2"] = "xliff2";
    Format["Xmb"] = "xmb";
})(Format || (exports.Format = Format = {}));
/**
 * How to handle duplicate translations.
 */
var I18NDuplicateTranslation;
(function (I18NDuplicateTranslation) {
    I18NDuplicateTranslation["Error"] = "error";
    I18NDuplicateTranslation["Ignore"] = "ignore";
    I18NDuplicateTranslation["Warning"] = "warning";
})(I18NDuplicateTranslation || (exports.I18NDuplicateTranslation = I18NDuplicateTranslation = {}));
